src/PolyType.SourceGenerator/Parser/Parser.cs

```diff
@@ -20,31 +20,96 @@ public sealed partial class Parser : TypeDataModelGenerator
         CommonHelpers.CreateTupleComparer<ITypeSymbol, string>(
             SymbolEqualityComparer.Default,
             CommonHelpers.CamelCaseInvariantComparer.Instance);
-    
+
     private readonly PolyTypeKnownSymbols _knownSymbols;
+    private readonly ImmutableDictionary<INamedTypeSymbol, TypeExtensionModel> _typeShapeExtensions;
 
-    private Parser(ISymbol generationScope, PolyTypeKnownSymbols knownSymbols, CancellationToken cancellationToken)
+    private Parser(ISymbol generationScope, ImmutableDictionary<INamedTypeSymbol, TypeExtensionModel> typeShapeExtensions, PolyTypeKnownSymbols knownSymbols, CancellationToken cancellationToken)
         : base(generationScope, knownSymbols, cancellationToken)
     {
         _knownSymbols = knownSymbols;
+        _typeShapeExtensions = typeShapeExtensions;
     }
 
     public static TypeShapeProviderModel? ParseFromGenerateShapeAttributes(
         ImmutableArray<TypeWithAttributeDeclarationContext> generateShapeDeclarations,
         PolyTypeKnownSymbols knownSymbols,
+        ImmutableDictionary<INamedTypeSymbol, TypeExtensionModel> typeShapeExtensions,
         CancellationToken cancellationToken)
     {
         if (generateShapeDeclarations.IsEmpty)
         {
             return null;
         }
 
-        Parser parser = new(knownSymbols.Compilation.Assembly, knownSymbols, cancellationToken);
+        Parser parser = new(knownSymbols.Compilation.Assembly, typeShapeExtensions, knownSymbols, cancellationToken);
         TypeDeclarationModel shapeProviderDeclaration = CreateShapeProviderDeclaration(knownSymbols.Compilation);
         ImmutableEquatableArray<TypeDeclarationModel> generateShapeTypes = parser.IncludeTypesUsingGenerateShapeAttributes(generateShapeDeclarations);
         return parser.ExportTypeShapeProviderModel(shapeProviderDeclaration, generateShapeTypes);
     }
 
+    public static ImmutableArray<TypeExtensionModel> DiscoverTypeShapeExtensions(KnownSymbols knownSymbols, Compilation compilation, MetadataReference metadataReference, CancellationToken cancellationToken)
```

Comment:
Should this be private?

Reply:
No. It's used in PolyTypeGenerator. It could be `internal`, but I don't see you using that very much in these projects, so I was trying to match your style.